src/vision/deliverables/borrowed_data_and_cancellation.md

```diff
@@ -0,0 +1,117 @@
+
+* [capability](./scopes/capability.md)
+* [APIs](./scopes/api.md)
+
+## Design notes
+
+Async functions are commonly written with borrowed references as arguments:
+
+```rust
+async fn do_something(db: &Db) { ... }
+```
+
+but important utilities like `spawn` and `spawn_blocking` require `'static` tasks. Building on non-cancelable traits, we can implement a "scope" API that allows one to introduce an async scope. This scope API should permit one to spawn tasks into a scope, but have various kinds of scopes (e.g., synchronous execution, parallel execution, and so forth). It should ultimately reside in the standard library and hook into different runtimes for scheduling. This will take some experimentation!
+
+```rust
+async fn foo(db: &Database) {
+    let result = std::async_thread::scope(|s| {
+        let job1 = s.spawn(async || {
+            async_thing(db)
+        });
+        let job2 = s.spawn_blocking(|| {
+            sync_thing(db)
+        });
+
+        (job1.await, job2.await)
+    }).await;
+}
+```
+
+### Side-stepping the nested await problem
+
+One goal of scopes is to avoid the "nested await" problem, as described in [Barbara battles buffered streams (BBBS)][bbbs]. The idea is like this: the standard combinators which run work "in the background" and which give access to intermediate results from that work should schedule that work into a scope.[^hard] This would typically be done by using an "interior iterator" pattern, but it could also be done by taking a scope parameter. Some examples from today's APIs are `FuturesUnordered` and `Stream::buffered`.
+
+[^hard]: This is not a hard rule. But invoking poll manually is best regarded as a risky thing to be managed with care -- not only because of the formal safety guarantees, but because of the possibility for "nested await"-style failures.
+
+[bbbs]: https://rust-lang.github.io/wg-async-foundations/vision/status_quo/barbara_battles_buffered_streams.html
+[`buffered`]: https://docs.rs/futures/0.3.15/futures/prelude/stream/trait.StreamExt.html#method.buffered
+
+In the case of [BBBS], the problem arises because of `buffered`, which spawns off concurrent work to process multiple connections. Under this system, the implementation of `buffered` would create an internal scope for spawn its tasks into that scope, side-stepping the problem. One could imagine also offering a variant of `buffered` like `buffered_in` that takes a scope parameter, permitting the user to choose the scope of those spawned tasks:
+
+```rust
+async fn do_work(database: &Database) {
+    std::async_thread::scope(|s| {
+        let work = do_select(database, FIND_WORK_QUERY)?;
+        std::async_iter::from_iter(work)
+            .map(|item| do_select(database, work_from_item(item)))
+            .buffered_in(5, scope)
+            .for_each(|work_item| process_work_item(database, work_item))
+            .await;
+    }).await;
+}
+```
+
+### Concurrency without scopes: Join, select, race, and friends
```

Comment:
I thought about that, but I don't think that's correct. I don't think it's correct that we should be concurrently fetching the next item for all async iterators while also processing the current one. It might be what you want sometimes, but I think if so, you should ask for it, which is precisely what I'm proposing here =)